README.md
## moveit_python

This is a set of pure-python bindings to the ROS interface of MoveIt! based on the
earlier moveit_utils package that was developed as part of the chess_player package.

## Overview

There are three interfaces currently:

 * MoveGroupInterface -- used to move the arm using the move_group action.
 * PlanningSceneInterface -- used to add/remove collision and attached objects.
   Can also set the colors of objects.
 * PickPlaceInterface -- used to interface to the pick and place actions.

## MoveGroupInterface

The MoveGroupInterface is quite easy to use:

```python
import rospy
from moveit_python import *

rospy.init_node("moveit_py")
g = MoveGroupInterface("planning_group_name", "base_link")
```

Obviously, you might need different values for base_link, and your planning group
is probably not called "planning_group_name".

## PlanningSceneInterface

The PlanningSceneInterface allows you to easily insert and remove objects from
the MoveIt! planning scene. Starting with version 0.3.0, this module will try
to use the newer service-based approach to apply planning scene updates, as
this is much more robust than publishing messages over topics asynchronously.

```python
import rospy
from moveit_python import *

rospy.init_node("moveit_py")
# create a planning scene interface, provide name of root link
p = PlanningSceneInterface("base_link")

# add a cube of 0.1m size, at [1, 0, 0.5] in the base_link frame
p.addCube("my_cube", 0.1, 1, 0, 0.5)

# do something

# remove the cube
p.removeCollisionObject("my_cube")
```

If for some reason you would prefer to not use the service, simply set
"use_service" to false in the various add/remove calls. Calling waitForSync
will republish messages as needed to synchronize the planning scene between
your script and MoveIt.

```python
p.addCube("my_cube", 0.1, 1, 0, 0.5, use_service=False)
p.addCube("my_other_cube", 0.1, 2, 0, 0.5, use_service=False)
p.waitForSync()
```

## PickPlaceInterface

```python
import rospy
from moveit_python import *
from moveit_msgs.msg import Grasp, PlaceLocation

rospy.init_node("moveit_py")
# provide arm group and gripper group names
# also takes a third parameter "plan_only" which defaults to False
p = PickPlaceInterface("arm", "gripper")

g = Grasp()
# fill in g
# setup object named object_name using PlanningSceneInterface
p.pickup("object_name", [g, ], support_name = "supporting_surface")

l = PlaceLocation()
# fill in l
p.place("object_name" [l, ], goal_is_eef = True, support_name = "supporting_surface")
```

## Migration from moveit_utils

 * GraspingInterface renamed to PickPlaceInterface
   * The pick/place functions now return the entire action result, in moveit_utils they returned only the error_code. To access the error code as it used to be returned, you would use result.error_code.val
 * ObjectManager renamed to PlanningSceneInterface
   * remove function is now removeCollisionObject and removeAttachedObject
 * ArmInterface renamed to MoveGroupInterface


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package moveit_python
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.4.5 (2023-01-03)
------------------
* fix CI after adding setup tools dependency
* Contributors: Michael Ferguson

0.4.4 (2022-12-23)
------------------
* use setuptools (`#33 <https://github.com/mikeferguson/moveit_python/issues/33>`_)
  setup from distutils is deprecated and will be removed eventually.
  It already breaks on Debian testing.
* Contributors: Michael Görner

0.4.3 (2022-01-14)
------------------
* fix: Add `attached_object_touch_links` to pickup (`#31 <https://github.com/mikeferguson/moveit_python/issues/31>`_)
* Contributors: Kai Waelti

0.4.2 (2021-06-28)
------------------
* New cone primitive and box primitive frame fix (`#30 <https://github.com/mikeferguson/moveit_python/issues/30>`_)
  * Box primitive frame independent of planning scene
  * Added cone solid primitive
* Contributors: Kiran Prasad

0.4.1 (2021-02-11)
------------------
* add support for objects not in fixed_frame
* Contributors: Michael Ferguson

0.4.0 (2021-01-17)
------------------
* various fixes for noetic/python3 (`#28 <https://github.com/mikeferguson/moveit_python/issues/28>`_)
* Contributors: Michael Ferguson

0.3.6 (2020-12-25)
------------------
* update package.xml for python3 support
* Contributors: Michael Ferguson

0.3.5 (2020-12-25)
------------------
* use "is None"
* Fixed a comparison and logical statement order issue. (`#26 <https://github.com/mikeferguson/moveit_python/issues/26>`_)
* fix pyassimp indices bug (`#27 <https://github.com/mikeferguson/moveit_python/issues/27>`_)
* Contributors: Karl Kangur, Michael Ferguson, Shingo Kitagawa

0.3.4 (2020-04-22)
------------------
* Fixed error message when removing attached object. (`#25 <https://github.com/mikeferguson/moveit_python/issues/25>`_)
  Co-authored-by: Karl Kangur <karl.kangur@exwzd.com>
* Merge pull request `#24 <https://github.com/mikeferguson/moveit_python/issues/24>`_ from v4hn/pr-master-python3
  python3 compatible API & syntax
* python3 compatible API & syntax
* Contributors: Karl Kangur, Michael Ferguson, v4hn

0.3.3 (2019-11-21)
------------------
* add sphere to planning scene interface
* add move_group parameter to MoveGroup class
* Contributors: Michael Ferguson, arwtyxouymz

0.3.2 (2019-11-14)
------------------
* enable pyassimp for melodic
* update maintainer email
* Contributors: Michael Ferguson, Shingo Kitagawa

0.3.1 (2018-08-13)
------------------
* add missing namespace to apply service
* Contributors: Michael Ferguson

0.3.0 (2018-07-31)
------------------
* add namespace functionality on planning_scene_interface.py
* additional cleanup/documentation
  * use apply_service for colors
  * rename sentUpdate to sendUpdate, add docs
  * rename wait param to use_service (the meaning has changed)
  * remove some spammy logging
  * don't waitForSync when using service
* Merge pull request `#10 <https://github.com/mikeferguson/moveit_python/issues/10>`_ from alemme/master
  adapt the code for the `apply` service
* added services to add objects to environment and attach them. Following http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/pr2_tutorials/planning/src/doc/planning_scene_ros_api_tutorial.html#interlude-synchronous-vs-asynchronous-updates
* Contributors: Benjamin-Tan, Lemme, Michael Ferguson

0.2.17 (2016-08-23)
-------------------
* Merge pull request `#9 <https://github.com/mikeferguson/moveit_python/issues/9>`_ from mikeferguson/pyassimp_fix
  pyassimp is broken in 16.04, temporary work around so we can release
* Contributors: Michael Ferguson

0.2.16 (2015-11-02)
-------------------
* fix error message to have proper function name
* Contributors: Michael Ferguson

0.2.15 (2015-08-20)
-------------------
* Fix types in move_group_interface
* add planning_scene_diff as optional field
* Contributors: Michael Ferguson, Mikkel Rath Pedersen

0.2.14 (2015-05-22)
-------------------
* add no-wait behavior for move, pick, and place
* updates for compliance with PEP8
* Contributors: Aaron Blasdel, Michael Ferguson

0.2.13 (2015-04-18)
-------------------
* better handle removal of objects
* place has no attached_object_touch_links
* Contributors: Michael Ferguson

0.2.12 (2015-04-11)
-------------------
* fixup planner id
* Contributors: Michael Ferguson

0.2.11 (2015-04-08)
-------------------
* fix spelling issue in velocity scaling factor
* Contributors: Michael Ferguson

0.2.10 (2015-04-06)
-------------------
* implement kwargs for pick&place interface
* add max_velocity_scaling_factor as kwarg
* allow overriding allowed_planning_time
* add FakeGroupInterface
* add clear() method to planning scene interface
* Contributors: Michael Ferguson

0.2.9 (2015-03-28)
------------------
* add rotate_pose_msg_about_origin
* Contributors: Michael Ferguson

0.2.8 (2015-03-21)
------------------
* expose num_attempts through kw_args
* Contributors: Michael Ferguson

0.2.7 (2014-11-19)
------------------
* enable removing attached objects
* Contributors: Michael Ferguson

0.2.6 (2014-11-16)
------------------
* use kw_args for group interface, add a number of args supported
* a couple of fixes for is_diff
* Contributors: Michael Ferguson

0.2.5 (2014-07-09)
------------------
* add scripts for dumping/loading planning scene
* documentation cleanup
* Contributors: Michael Ferguson

0.2.4 (2014-06-03)
------------------
* add list and remove object scripts, closes `#2 <https://github.com/mikeferguson/moveit_python/issues/2>`_
* properly initialize planning scene interface, fixes `#1 <https://github.com/mikeferguson/moveit_python/issues/1>`_
* add planner_id logic to move group interface
* remove default support name
* fix comment on result type
* add missing import
* upstream the retry logic
* Contributors: Michael Ferguson

0.2.3 (2014-05-26)
------------------
* fix bug in mesh generation
* Contributors: Michael Ferguson

0.2.2 (2014-05-21)
------------------
* pass full result in pick and place

src/moveit_python/pick_place_interface.py
# Copyright 2015, Fetch Robotics Inc
# Copyright 2011-2014, Michael Ferguson
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import rospy
import actionlib
from moveit_msgs.msg import MoveItErrorCodes
from moveit_msgs.msg import PickupAction, PickupGoal, PlaceAction, PlaceGoal

## @brief Simple interface to pick and place actions
class PickPlaceInterface(object):

    ## @brief Create a grasp manager, connect actions
    ## @param group Name of arm planning group
    ## @param ee_group Name of end effector planning group
    ## @param plan_only Should we only plan, but not execute?
    def __init__(self,
                 group="arm",
                 ee_group="gripper",
                 plan_only=False,
                 verbose=False):
        self._verbose = verbose
        self._group = group
        self._effector = ee_group
        if self._verbose:
            rospy.loginfo("Connecting to pickup action...")
        self._pick_action = actionlib.SimpleActionClient("pickup",
                                                         PickupAction)
        self._pick_action.wait_for_server()
        if self._verbose:
            rospy.loginfo("...connected")
            rospy.loginfo("Connecting to place action...")
        self._place_action = actionlib.SimpleActionClient("place",
                                                          PlaceAction)
        self._place_action.wait_for_server()
        if self._verbose:
            rospy.loginfo("...connected")
        self._plan_only = plan_only
        self.planner_id = None
        self.allowed_planning_time = 30.0

    def get_pick_action(self):
        return self._pick_action

    def get_place_action(self):
        return self._place_action

    ## @brief Plan and grasp something
    ## @param name Name of the object to grasp
    ## @param grasps Grasps to try (moveit_msgs/Grasp)
    ## @param support_name Name of the support surface
    ## @returns moveit_msgs/PickupResult
    def pickup(self, name, grasps, wait=True, **kwargs):
        # Check arguments
        supported_args = ("allow_gripper_support_collision",
                          "attached_object_touch_links",
                          "allowed_touch_objects",
                          "plan_only",
                          "planner_id",
                          "planning_scene_diff",
                          "planning_time",
                          "support_name")
        for arg in kwargs.keys():
            if not arg in supported_args:
                rospy.loginfo("pickup: unsupported argument: %s", arg)

        # Create goal
        g = PickupGoal()

        # 1. Fill in target name
        g.target_name = name

        # 2. Fill in group name
        g.group_name = self._group

        # 3. Fill in end effector
        g.end_effector = self._effector

        # 4. List of grasps
        g.possible_grasps = grasps

        # 5. Fill in support surface
        try:
            g.support_surface_name = kwargs["support_name"]
        except KeyError:
            g.support_surface_name = ""

        # 6. Can gripper contact the support surface (named above)
        try:
            g.allow_gripper_support_collision = kwargs["allow_gripper_support_collision"]
        except KeyError:
            g.allow_gripper_support_collision = True

        # 7. What links are part of the "gripper" (and can contact the object/surface)
        try:
            g.attached_object_touch_links = kwargs["attached_object_touch_links"]
        except KeyError:
            g.attached_object_touch_links = list() # empty list = use all links of end-effector

        # 8. Fill in minimize_object_distance

        # 9. Fill in path_constraints

        # 10. Fill in planner id
        try:
            g.planner_id = kwargs["planner_id"]
        except KeyError:
            if self.planner_id:
                g.planner_id = self.planner_id

        # 11. List of obstacles that we are allowed to touch
        try:
            g.allowed_touch_objects = kwargs["allowed_touch_objects"]
        except KeyError:
            g.allowed_touch_objects = list()

        # 12. Planning time
        try:
            g.allowed_planning_time = kwargs["planning_time"]
        except KeyError:
            g.allowed_planning_time = self.allowed_planning_time

        # 13. Planning options
        try:
            g.planning_options.planning_scene_diff = kwargs["planning_scene_diff"]
        except KeyError:
            g.planning_options.planning_scene_diff.is_diff = True
            g.planning_options.planning_scene_diff.robot_state.is_diff = True
        g.planning_options.plan_only = self._plan_only

        self._pick_action.send_goal(g)
        if wait:
            self._pick_action.wait_for_result()
            return self._pick_action.get_result()
        else:
            return None

    ## @brief Plan and grasp something
    ## @param name Name of the object to grasp
    ## @param grasps Grasps to try (moveit_msgs/Grasp)
    ## @param support_name Name of the support surface
    ## @param goal_is_eef Set to true if the place goal is for the
    ##        end effector frame, default is object frame.
    ## @returns moveit_msgs/PlaceResult
    def place(self, name, locations, wait=True, **kwargs):
        # Check arguments
        supported_args = ("allow_gripper_support_collision",
                          "allowed_touch_objects",
                          "goal_is_eef",
                          "plan_only",
                          "planner_id",
                          "planning_scene_diff",
                          "planning_time",
                          "support_name")
        for arg in kwargs.keys():
            if not arg in supported_args:
                rospy.loginfo("place: unsupported argument: %s", arg)

        # Create goal
        g = PlaceGoal()

        # 1. Name of ARM planning group
        g.group_name = self._group

        # 2. Name of attached object to place
        g.attached_object_name = name

        # 3. Possible locations to place
        g.place_locations = locations

        # 4. If true, using eef pose (same as pick)
        try:
            g.place_eef = kwargs["goal_is_eef"]
        except KeyError:
            g.place_eef = False

        # 5. Fill in support surface
        try:
            g.support_surface_name = kwargs["support_name"]
        except KeyError:
            g.support_surface_name = ""

        # 6. Can gripper contact the support surface (named above)
        try:
            g.allow_gripper_support_collision = kwargs["allow_gripper_support_collision"]
        except KeyError:
            g.allow_gripper_support_collision = True

        # 8. Fill in path_constraints

        # 9. Fill in planner id
        try:
            g.planner_id = kwargs["planner_id"]
        except KeyError:
            if self.planner_id:
                g.planner_id = self.planner_id

        # 10. List of obstacles that we are allowed to touch
        try:
            g.allowed_touch_objects = kwargs["allowed_touch_objects"]
        except KeyError:
            g.allowed_touch_objects = list()

        # 11. Planning time
        try:
            g.allowed_planning_time = kwargs["planning_time"]
        except KeyError:
            g.allowed_planning_time = self.allowed_planning_time

        # 12. Planning options
        try:
            g.planning_options.planning_scene_diff = kwargs["planning_scene_diff"]
        except KeyError:
            g.planning_options.planning_scene_diff.is_diff = True
            g.planning_options.planning_scene_diff.robot_state.is_diff = True
        g.planning_options.plan_only = self._plan_only

        self._place_action.send_goal(g)
        if wait:
            self._place_action.wait_for_result()
            return self._place_action.get_result()
        else:
            return None

    ## Common usage pattern
    ## TODO document
    def pick_with_retry(self, name, grasps, retries=5, scene=None, **kwargs):
        if self._verbose:
            rospy.loginfo("Beginning to pick.")
        while retries > 0:
            retries += -1
            pick_result = self.pickup(name, grasps, **kwargs)
            if pick_result.error_code.val == MoveItErrorCodes.SUCCESS:
                rospy.loginfo("Pick succeeded")
                return [True, pick_result]
            elif pick_result.error_code.val == MoveItErrorCodes.PLANNING_FAILED:
                rospy.logerr("Pick failed in the planning stage, try again...")
                rospy.sleep(0.5)  # short sleep to try and let state settle a bit?
                continue
            elif not scene is None and \
                (pick_result.error_code.val == MoveItErrorCodes.CONTROL_FAILED or \
                pick_result.error_code.val == MoveItErrorCodes.MOTION_PLAN_INVALIDATED_BY_ENVIRONMENT_CHANGE or \
                pick_result.error_code.val == MoveItErrorCodes.TIMED_OUT):
                rospy.logerr("Pick failed during execution, attempting to cleanup.")
                if name in scene.getKnownAttachedObjects():
                    rospy.loginfo("Pick managed to grab object, retreat must have failed, continuing anyways")
                    return [True, pick_result]
                else:
                    rospy.loginfo("Pick did not grab object, try again...")
                    continue
            else:
                rospy.logerr("Pick failed with error code: %d. Will retry...",
                             pick_result.error_code.val)
                continue
        rospy.logerr("Pick failed, and all retries are used")
        return [False, pick_result]

    ## Common usage pattern
    ## TODO document
    def place_with_retry(self,
                         name,
                         locations,
                         retries=5,
                         scene=None,
                         **kwargs):
        if self._verbose:
            rospy.loginfo("Beginning to place.")
        while retries > 0:
            retries += -1
            place_result = self.place(name, locations, **kwargs)
            if place_result.error_code.val == MoveItErrorCodes.SUCCESS:
                rospy.loginfo("Place succeeded")
                return [True, place_result]
            elif place_result.error_code.val == MoveItErrorCodes.PLANNING_FAILED:
                rospy.logerr("Place failed in planning stage, try again...")
                rospy.sleep(0.5)  # short sleep to let state settle a bit?
                continue
            elif not scene is None and \
                 (place_result.error_code.val == MoveItErrorCodes.CONTROL_FAILED or \
                 place_result.error_code.val == MoveItErrorCodes.MOTION_PLAN_INVALIDATED_BY_ENVIRONMENT_CHANGE or \
                 place_result.error_code.val == MoveItErrorCodes.TIMED_OUT):
                rospy.logerr("Place failed during execution, attempting to cleanup.")
                if name in scene.getKnownAttachedObjects():
                    rospy.loginfo("Place did not place object, approach must have failed, will retry...")
                    continue
                else:
                    rospy.loginfo("Object no longer in gripper, must be placed, continuing...")
                    return [True, place_result]
            else:
                rospy.logerr("Place failed with error code: %d. Will retry...",
                             place_result.error_code.val)
                continue
        rospy.logerr("Place failed, and all retries are used")
        return [False, place_result]
